### DISCUSSION TITLE
should result props rather be specifically typed to avoid type collissions

### DISCUSSION THREAD

**Discussion by AndyOGo** [adobe/react-spectrum#1618] [General]
Im using `useKeyboard` which returns `{ keyboardProps }`.
These `keyboardProps` are typed too generically with `HTMLAttributes<HTMLElement>`, because they only set `onKeyDown` and `onKeyUp` and all the other possibly collide if a component allows other types.

https://github.com/adobe/react-spectrum/blob/f6c06605243ad2033fce95f80ae3fecd4a38daeb/packages/%40react-aria/interactions/src/useKeyboard.ts#L22-L37

I think it's great to type the input of the hooks generically, but the output should be typed specifically to avoid type collisions with custom made components.

Typescript's [utility type `Pick<T, Keys>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys) could be utilized to improve these typings, like:

```ts
interface KeyboardResult {
  /** Props to spread onto the target element. */
  keyboardProps: Pick<HTMLAttributes<HTMLElement>, "onKeyDown" | "onKeyUp">
}
```

Otherwise I get this TS error for my custom `onChange` event:

```sh
Type '{ className: string; type: string; value: string; renderLeftSide: (() => Element) | undefined; ref: (node: any) => void; defaultChecked?: boolean | undefined; defaultValue?: string | number | readonly string[] | undefined; ... 280 more ...; labelProps: { ...; }; }' is not assignable to type 'Pick<InputProps, "onChange" | "accept" | "acceptCharset" | "action" | "allowFullScreen" | "allowTransparency" | "alt" | "as" | "async" | "autoComplete" | "autoFocus" | ... 353 more ... | "isInvalid">'.
  Types of property 'onChange' are incompatible.
    Type '(((event: FormEvent<HTMLElement>) => void) & ((event: ChangeEvent<HTMLInputElement & HTMLTextAreaElement>) => void)) | undefined' is not assignable to type '((value: string) => void) | undefined'.
      Type '((event: FormEvent<HTMLElement>) => void) & ((event: ChangeEvent<HTMLInputElement & HTMLTextAreaElement>) => void)' is not assignable to type '((value: string) => void) | undefined'.
        Type '((event: FormEvent<HTMLElement>) => void) & ((event: ChangeEvent<HTMLInputElement & HTMLTextAreaElement>) => void)' is not assignable to type '(value: string) => void'.
          Types of parameters 'event' and 'value' are incompatible.
            Type 'string' is not assignable to type 'FormEvent<HTMLElement>'.ts(2322)
```

Any hook result with a generic type would need a refactoring:
https://github.com/adobe/react-spectrum/search?q=HTMLAttributes%3CHTMLElement%3E

**Comment by AndyOGo** [comment]
I started a PR for the integration hooks
https://github.com/adobe/react-spectrum/pull/1619/files

**Comment by devongovett** [comment]
The returned props are intentionally very broad. You should treat them as opaque implementation details and spread them directly onto DOM elements or pass them around as a whole rather than individually. This allows us to change the implementation in the future to return more or less props without breaking changes. If we typed each prop individually, then we could never remove props, and adding props would also potentially break things if only the previously used props were passed around.

I'm curious where you're seeing TS errors. It makes me think the returned props are being used somewhere other than being spread on a DOM element, which seems like it would be problematic.

**Reply by AndyOGo** [reply]
Thank you for your quick reply.
I see the intention behind it.

Indeed, that is it. The props are spread onto an `<Input>` abstraction and there it starts where this generic approach hinders compatibility.

This example `<Input>` abstracts the `onChange` event, like:

`Input.tsx`:
```tsx
import React, { useCallback } from "react";

export interface InputProps extens Omit<React.AllHTMLAttributes<HTMLInputElement>, "onChange"> {
  onChange: (value: string) => void; // I could also rename this to onChangeValue...
}

export function Input({ onChange, ...props}: InputProps) {
  const handleChange = useCallback((event) => {
    if(onChange)
      onChange(event.target.value)
    }
  })
  return <input {...props} onChange={handleChange} />;
}
```

Then I use this `<Input>` component with `useKeyboard`.

`AnotherInput.tsx`:
```tsx
import {useKeyboard} from '@react-aria/interactions'
import { Input } from "./Input";

function AnotherInput() {
  // for brevity I omit the details here...
  const { keyboardProps } = useKeyboard({ ... });
  
  // here is the problem
  // the type of never used keyboardProps onChange collides with the one from Input
  return <Input {...keyboardProps} />
}

**Reply by AndyOGo** [reply]
Im curios if it would help to apply broad types only if needed?

**Reply by AndyOGo** [reply]
In fact `useTextfield` applies the same `onChange` simplification with the `ValueBase` type.

https://github.com/adobe/react-spectrum/blob/fa314396b5ba0d20667362ad1c1f1fc4aee59ccd/packages/%40react-aria/textfield/src/useTextField.ts#L80

https://github.com/adobe/react-spectrum/blob/fa314396b5ba0d20667362ad1c1f1fc4aee59ccd/packages/%40react-types/shared/src/inputs.d.ts#L32-L39

**Reply by devongovett** [reply]
Yeah, this can happen sometimes. I see two possibilities, which we use in React Spectrum itself to handle this in some places.

1. Add a prop that accepts DOM props.

```tsx
<Input domProps={inputProps} />
```

2. Use a context to pass DOM props:

```tsx
<DOMPropsProvider value={inputProps}>
  <Input />
</DOMPropsProvider>
```

And make `Input` consume from that context.

In your case, the `onChange` prop returned from `useTextField` actually doesn't match your `Input` component's props, so the TS error is probably correct. You need to either find a way to pass the DOM props through, or adapt them to match your component's API.

**Reply by AndyOGo** [reply]
Thanks for your nice reply @devongovett 

Meanwhile I moved on with option 1 and I layered our `onChange` simplification one level up.

From the viewpoint of a library consumer I still think the typings are too broad, especially if only 2 or 3 out of many dozen DOM props are actually being used.

Anyway this is a great library and I'm sure you considered much more use cases than I have to deal with right now.

**Reply by venikx** [reply]
Would this be okay, while keeping the types as wide as possible? https://github.com/adobe/react-spectrum/pull/3287

**Comment by venikx** [comment]
@devongovett 
I might be able to chime in here, because I ran into the same issues in an application I'm working on, when I pulled down react-spectrum and re-implemented the component using the libs from react-spectrum I did not run into the issue. For me specially there's a type mismatch when using all the hooks from the useDialog example. 

I traced it back to the following:
From my app
```ts
    interface FocusEvent<T = Element> extends SyntheticEvent<T, NativeFocusEvent> {
        relatedTarget: EventTarget | null;
        target: EventTarget & T;
    }
```

From react-spectrum
```ts
    interface FocusEvent<Target = Element, RelatedTarget = Element> extends SyntheticEvent<Target, NativeFocusEvent> {
        relatedTarget: (EventTarget & RelatedTarget) | null;
        target: EventTarget & Target;
    }
```

So I wonder, if I run a different version of @types/react than react-spectrum, would typings be slightly different?
